src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,150 @@
+
+            // Can't use a UTF8 string inside an expression tree.
+            var semanticModel = context.Operation.SemanticModel;
+            Contract.ThrowIfNull(semanticModel);
+            if (arrayCreationExpression.Syntax.IsInExpressionTree(semanticModel, expressionType, context.CancellationToken))
+                return;
+
+            var values = arrayCreationExpression.Initializer.ElementValues.SelectAsArray(v => v.ConstantValue.HasValue && v.ConstantValue.Value is not null, v => v.ConstantValue.Value);
+
+            // If we couldn't get constant values for all elements then we can't offer
+            if (values.Length != arrayCreationExpression.Initializer.ElementValues.Length)
+                return;
+
+            var stringValue = GetStringValue(values);
+
+            if (stringValue is null)
+                return;
+
+            var properties = ImmutableDictionary<string, string?>.Empty.Add(StringValuePropertyName, stringValue);
+
+            context.ReportDiagnostic(
+                DiagnosticHelper.Create(Descriptor, arrayCreationExpression.Syntax.GetLocation(), option.Notification.Severity, additionalLocations: null, properties));
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        private static string? GetStringValue(ImmutableArray<object?> values)
+        {
+            try
+            {
+                var byteValues = new byte[values.Length];
+                for (var i = 0; i < values.Length; i++)
+                {
+                    // We shouldn't get nulls, but Convert.ToByte will return (byte)0 if there are any
+                    // so better to crash if the analyzer has a bug, that output a buggy string in users code.
+                    if (values[i] is null)
+                        throw ExceptionUtilities.Unreachable;
+
+                    byteValues[i] = Convert.ToByte(values[i]);
+                }
+
+                var pooledBuilder = PooledStringBuilder.GetInstance();
+                var builder = pooledBuilder.Builder;
+                var ros = new ReadOnlySpan<byte>(byteValues);
+                for (var i = 0; i < ros.Length;)
+                {
+                    // If we can't decode a rune from the array then it can't be represented as a string
+                    if (Rune.DecodeFromUtf8(ros.Slice(i), out var rune, out var bytesConsumed) != OperationStatus.Done)
+                        return null;
+
+                    i += bytesConsumed;
+                    builder.Append(GetStringLiteralRepresentation(rune));
+                }
+
+                return pooledBuilder.ToStringAndFree();
+            }
+            catch
+            {
+                // Ignore any conversion failures and just don't report the diagnostic
+            }
+
+            return null;
+        }
+
+        private static string GetStringLiteralRepresentation(Rune rune)
```

Comment:
https://github.com/dotnet/roslyn/blob/218232bd168b9fcad8a5a825b1b1375398ccd413/src/Compilers/CSharp/Portable/SymbolDisplay/ObjectDisplay.cs#L144-L190

This is a similar method (not `public` from the compiler layer, so it cannot be reused). But it does more work:

```csharp
     if (NeedsEscaping(CharUnicodeInfo.GetUnicodeCategory(c))) 
     { 
         replaceWith = "\\u" + ((int)c).ToString("x4"); 
         return true; 
     } 
```

So wondering if the extra work is necessary for this feature?

Reply:
i do think there is merit in using either \u or \U for things like control/formatting characters.  Those will likely show up very weird once in string-literal form.

Reply:
I deliberately didnt do anything (yet?) about `\u` because it's not that simple. My HalfEmoji test is a good example. That byte array should be `"\uD83D"`, but whilst that is a valid string literal, `"\uD83D"u8` is not a valid UTF8 string literal.

Reply:
Also worth mentioning similar code to this exists in IVirtualCharService.TryGetEscapeCharacter but it was slightly painful to use. We could definitely expand that system if we wanted to (allow it to take in strings, not just syntax, and produce escaped code as well as consuming it) and use it here instead of direct Rune usage potentially